pkg/services/hcloud/server/server.go

```diff
@@ -189,6 +189,22 @@ func (s *Service) Reconcile(ctx context.Context) (res reconcile.Result, err erro
 	return res, nil
 }
 
+// implments setting rate limit  on hcloudmachine.
+func handleRateLimit(hm *infrav1.HCloudMachine, err error, functionName string, errMsg string) error {
+	// check for a rate limit exceeded error if the machine is not running or if machine has a deletion timestamp
+	if !hm.Status.Ready || !hm.DeletionTimestamp.IsZero() {
+		hcloudutil.HandleRateLimitExceeded(hm, err, functionName)
+		return fmt.Errorf("%s: %w", errMsg, err)
+	}
+
+	// returns nil for a rate limit exceeded error
+	if hcloud.IsError(err, hcloud.ErrorCodeRateLimitExceeded) {
+		return nil
+	}
+
+	return fmt.Errorf("%s: %w", errMsg, err)
+}
```

Comment:
I thought about how this could be changed a bit to make it easier to understand. Here is untested pseudo code  (using "return early" pattern):

```go
	if !hcloud.IsError(err, hcloud.ErrorCodeRateLimitExceeded) {
        // no rate-limit error. Don't ignore those
		return fmt.Errorf("%s: %w", errMsg, err)
	}
    if hm.Status.Ready && hm.DeletionTimestamp.IsZero() {
        // The machine is up and running. Ignore the rate-limit for this machine.
        return nil
    }

    // The machine is in-progress. Let the user know about the rate-limit
    hcloudutil.HandleRateLimitExceeded(hm, err, functionName)
    return fmt.Errorf("%s: %w", errMsg, err)
```

@yrs147 what do you think about that?